src/tosyntax/FStar.ToSyntax.Resugar.fs

```diff
@@ -0,0 +1,614 @@
+open FStar.Syntax.Print
+module I = FStar.Ident
+module S  = FStar.Syntax.Syntax
+module SS = FStar.Syntax.Subst
+module A  = FStar.Parser.AST
+module C = FStar.Syntax.Const
+module U = FStar.Syntax.Util
+module BU = FStar.Util
+
+let bv_as_unique_ident (x:S.bv) : I.ident =
+  let unique_name =  x.ppname.idText ^ "__" ^ (string_of_int x.index) in
+  I.mk_ident (unique_name, x.ppname.idRange)
+
+let resugar_arg_qual (q:option<S.arg_qualifier>) : option<A.arg_qualifier> =
+  match q with
+    | None -> None
+    | Some (Implicit b) -> Some A.Implicit  //TODO: how should we map this flag back to the surface?
+    | Some Equality -> Some A.Equality
+
+let rec universe_to_int n u =
+  match u with 
+    | U_succ u -> universe_to_int (n+1) u
+    | _ -> (n, u)
+
+let rec resugar_universe (u:S.universe) r: A.term =
+  let mk (a:A.term') r: A.term =
+      //augment `a` an Unknown level (the level is unimportant ... we should maybe remove it altogether)
+      A.mk_term a r A.Un
+  in
+  begin match u with 
+    | U_zero -> 
+      mk (A.Const(Const_int ("0", None))) r 
+
+    | U_succ _ ->
+      let (n, u) = universe_to_int 0 u in
+      begin match u with
+      | U_zero -> 
+        mk (A.Const(Const_int(string_of_int n, Some (Unsigned, Int16)))) r 
+      
+      | _ -> 
+        let e1 = mk (A.Const(Const_int(string_of_int n, Some (Unsigned, Int16)))) r in
+        let e2 = resugar_universe u r in
+        mk (A.Op("+", [e1; e2])) r 
+      end
+
+    | U_max([]) -> failwith "Impossible: U_max without arguments"
+
+    | U_max (_1::_2) ->
+      let rec aux e = function
+        | [] -> e
+        | hd::tl ->
+          let e = mk (A.App(e, resugar_universe hd r, A.Nothing)) r in
+          aux e tl 
+      in
+      let t = mk (A.Var(lid_of_path ["max"] r)) r in
+      let e = mk (A.App(t, resugar_universe _1 r, A.Nothing)) r in
+      aux e _2
+
+    | U_name u -> mk (A.Uvar(u)) r
+    | U_unif _ -> mk A.Wild r
+    | U_bvar _ -> failwith "U_bvar is impossible in resugar universe"
+    | U_unknown -> failwith "U_unknow is impossible in resugar universe"
+  end
+
+let string_to_op s =
+  let (|Prefix|_|) (p:string) (s:string) =
+    if s.StartsWith(p) then
+        Some(s.Substring(p.Length))
+    else
+        None
+  in 
+  let name_of_op = function
+    | "Amp" -> Some "&"
+    | "At" -> Some "@"
+    | "Plus" -> Some "+"
+    | "Minus" -> Some "-"
+    | "Subtraction" -> Some "-"
+    | "Slash" -> Some "/"
+    | "Less" -> Some "<"
+    | "Equals" -> Some "="
+    | "Greater" -> Some ">"
+    | "Underscore" -> Some "_"
+    | "Bar" -> Some "|"
+    | "Bang" -> Some "!" 
+    | "Hat" -> Some "^"
+    | "Percent" -> Some "%"
+    | "Star" -> Some "*"
+    | "Question" -> Some "?"
+    | "Colon" -> Some ":"
+    | _ -> None
+  in
+  match s with 
+  | "op_String_Assignment" -> Some ".[]<-"
+  | "op_Array_Assignment" -> Some ".()<-"
+  | "op_String_Access" -> Some ".[]"
+  | "op_Array_Access" -> Some ".()"
+  | Prefix "op__" op -> name_of_op op
```

Comment:
If you look at `compile_op` in `src/parser/FStar.Parser.AST.fs` the last case is doing a `String.concat` over a *list* of operations. So `^+` will be compiled to `op_Hat_Plus` for instance. Btw there seem to be only one `_` after `op`